src/js/component-listing.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+const repoListing = require('../../lib/repo-listing');
+
+/**
+ * Represents a filterable component listing.
+ */
+class ComponentListing {
+
+	/**
+	 * Class constructor.
+	 * @param {HTMLElement} [listingElement] - The listing element in the DOM.
+	 */
+	constructor(listingElement) {
+		this.listingElement = listingElement;
+		this.components = this.getComponents();
+		this.categoryElements = this.getCategoryElements();
+		document.addEventListener('o.filterFormUpdate', this.handleFilterFormUpdateEvent.bind(this));
+	}
+
+	/**
+	 * Handle the filter form update event.
+	 */
+	handleFilterFormUpdateEvent(event) {
+		const filter = event.detail;
+
+		// Perform the visibility marking
+		this.components.map(component => {
+			delete component.visible;
+			return component;
+		})
+		this.components = repoListing.markVisibilityBySearchTerm(this.components, filter.search);
+		this.components = repoListing.markVisibilityByType(this.components, {
+			imageset: filter.imageset,
+			module: filter.module,
+			service: filter.service
+		});
+		this.components = repoListing.markVisibilityByStatus(this.components, {
+			active: filter.active,
+			dead: filter.dead,
+			deprecated: filter.deprecated,
+			experimental: filter.experimental,
+			maintained: filter.maintained
+		});
+
+		// Set classes and attributes
+		for (const component of this.components) {
+			if (component.visible) {
+				component.element.removeAttribute('aria-hidden');
+				component.element.classList.remove('o-registry-ui__table-row--hidden');
+			} else {
+				component.element.setAttribute('aria-hidden', 'true');
+				component.element.classList.add('o-registry-ui__table-row--hidden');
+			}
+		};
+		for (const [categoryName, category] of Object.entries(repoListing.categorise(this.components))) {
+			if (category.visible) {
+				this.categoryElements[categoryName].removeAttribute('aria-hidden');
+				this.categoryElements[categoryName].classList.remove('o-registry-ui__table-row--hidden');
+			} else {
+				this.categoryElements[categoryName].setAttribute('aria-hidden', 'true');
+				this.categoryElements[categoryName].classList.add('o-registry-ui__table-row--hidden');
+			}
+		}
+
+	}
+
+	/**
+	 * Get the list of components.
+	 */
+	getComponents() {
+		const elements = this.listingElement.querySelectorAll('[data-o-component]');
+		return Array.from(elements).map(element => {
```

Comment:
if I recall correctly, the second argument in `Array.from` is automatically a `.map` ?

Reply:
YES! I keep forgetting :D will update